content/blog/2021-04-01-fuzzing-grub-pt2.markdown

```diff
@@ -0,0 +1,248 @@
+userspace tools and `grub-emu` built. Basically it set `AFL_DEFER_LIB` when
+building all the modules and turned it off when building the userspace tools
+and `grub-emu`.
+
+This worked and it's what I used to find most of my bugs. But I'd probably not
+recommend it, and I'm not sharing the source: it's extremely fragile and
+brittle, the hard way is more generally applicable, and the easy way is nicer.
+
+
+The easy way: adjusting linker flags
+------------------------------------
+
+After posting part 1 of this series, I had a fascinating twitter DM conversation
+with [@hackerschoice](https://twitter.com/hackerschoice), who pointed me to some
+new work that had been done in AFL++ between when I started and when I published
+part 1.
+
+AFL++ now has the ability to dynamically detect some duplicate symbols, allowing
+it to support plugins and modules better. This isn't directly applicable because
+we link all the modules in at build time, but in the conversation I was pointed
+to a linker flag which instructs the linker to ignore the symbol duplication
+rather than error out. This provides a significantly simpler way to instrument
+`grub-emu`, avoiding all the issues I'd previously been fighting so hard to
+address.
+
+So, with a modern AFL++, and the patch from [part 1](https://github.com/daxtens/grub/tree/fuzzing-pt2),
+you can sort out this entire process like this:
+
+```
+:::shell
+./bootstrap
+./configure --with-platform=emu CC=clang CXX=clang++ --disable-grub-emu-sdl
+make CC=/path/to/afl-clang-fast LDFLAGS="-Wl,--allow-multiple-definition"
+```
+
+Eventually it will error out, but `./grub-core/grub-emu` should be successfully
+built first.
+
+(Why not just build `grub-emu` directly? It gets built by `grub-core/Makefile`,
+but depends on a bunch of things made by the top-level makefile and doesn't
+express its dependencies well. So you can try to build all the things that you
+need separately and then `cd grub-core; make ...flags... grub-emu` if you want -
+but it's way more complicated to do it that way!)
+
+
+Going extra fast: `__AFL_INIT`
+==============================
+
+Now that we can compile with instrumentation, we can use `__AFL_INIT`. I'll
+leave the precise details of how this works to the AFL docs, but in short it
+allows us to do a bunch of early setup only once, and just fork the process
+after the setup is done.
+
+There's a patch that inserts a call to `__AFL_INIT` in the `grub-emu` start path
+in [my GitHub repo](https://github.com/daxtens/grub/tree/fuzzing-pt2).
+
+All up, this can lead to a 2x-3x speedup over the figures I saw in [part 1](/blog/2021/03/04/fuzzing-grub-part-1).
+In part 1 we saw around 244 executions per second at this point - now we're over
+500:
+
+![afl-fuzz fuzzing grub, showing fuzzing happening](/images/dja/grub-fuzzing-pt2.png)
+
+
+Finding more bugs with sanitisers
+=================================
+
+In theory, this also makes it really easy to compile with ASAN, which can often
```

Comment:
A link or a few extra words here explaining what ASAN is would be nice. Otherwise looks good to me!

Reply:
Thank you, that's helpful. I've added a paragraph and repushed.